assets/js/components/settings/SettingsApp.js

```diff
@@ -19,81 +19,32 @@
 /**
  * External dependencies
  */
+import PropTypes from 'prop-types';
 import Tab from '@material/react-tab';
 import TabBar from '@material/react-tab-bar';
-import { useFirstMountState, useHash } from 'react-use';
+import { withRouter, Link as NavLink } from 'react-router-dom';
 
 /**
  * WordPress dependencies
  */
-import { Fragment, useState } from '@wordpress/element';
+import { Fragment } from '@wordpress/element';
 import { __ } from '@wordpress/i18n';
 
 /**
  * Internal dependencies
  */
-import Data from 'googlesitekit-data';
-import SettingsAdmin from './SettingsAdmin';
 import Header from '../Header';
 import PageHeader from '../PageHeader';
 import Layout from '../layout/Layout';
 import HelpLink from '../HelpLink';
 import SettingsModules from './SettingsModules';
-import { CORE_MODULES } from '../../googlesitekit/modules/datastore/constants';
 import { Cell, Grid, Row } from '../../material-components';
 
-const { useSelect, useDispatch } = Data;
-
-const hashFrom = ( tabID, slug, state ) => {
-	const fragments = [ tabID ];
-
-	if ( tabID === 'settings' && slug && state !== 'closed' ) {
-		fragments.push( slug );
-		if ( state ) {
-			fragments.push( state );
-		}
-	}
-
-	return fragments.join( '/' );
-};
-
-const parseHash = ( hashToParse ) => hashToParse.replace( '#', '' ).split( /\// );
-
-export default function SettingsApp() {
-	const [ hash, setHash ] = useHash();
-	const isFirstMount = useFirstMountState();
-	const { setModuleSettingsPanelState } = useDispatch( CORE_MODULES );
-	const [ initialActiveTabID, initialModuleSlug, initialModuleState ] = parseHash( hash );
-	const [ activeTabID, setActiveTabID ] = useState( initialActiveTabID || 'settings' );
-	const [ moduleSlug, setModuleSlug ] = useState( initialModuleSlug );
-	const activeTab = SettingsApp.tabToIndex[ activeTabID ];
-
-	if ( isFirstMount ) {
-		if ( initialModuleSlug ) {
-			setModuleSettingsPanelState( initialModuleSlug, initialModuleState );
-		}
-		if ( ! hash ) {
-			setHash( 'settings' );
-		}
-	}
-
-	const moduleState = useSelect( ( select ) => moduleSlug ? select( CORE_MODULES ).getModuleSettingsPanelState( moduleSlug ) : null );
-
-	const setModuleState = ( slug, state ) => {
-		if ( state === 'edit' || state === 'view' ) {
-			setModuleSlug( slug );
-		} else {
-			setModuleSlug( null );
-		}
-		setModuleSettingsPanelState( slug, state );
-		setHash( hashFrom( activeTabID, slug, state ) );
-	};
-
-	const handleTabUpdate = ( tabIndex ) => {
-		const newActiveTabID = SettingsApp.tabIDsByIndex[ tabIndex ];
-		setActiveTabID( newActiveTabID );
-		setHash( hashFrom( newActiveTabID ) );
-	};
+function SettingsApp( { location: { pathname } } ) {
+	// Don't pass NavLink component if it would send you to same URL, prevents a warning.
+	const getTag = ( path ) => basePath === path ? 'button' : NavLink;
```

Comment:
Interesting, didn't know that would cause a warning. Sort of a shame!

Reply:
The specific warning being: `Warning: Hash history cannot PUSH the same path; a new entry will not be added to the history stack`

After looking at this again with a clearer head, I'm not a big fan of the `getTag` approach I went with, seems hacky.

New approach uses a `shouldReplaceHistory` method that is passed into the `replace` prop to `Tab` components, which is supplied to `NavLink` components. This seems a lot less hacky, we're replacing the history entry with an identical one rather than depending on the `button` element not responding to the `to` prop.